src/plugin/vk-wrapper/instance/Instance.cpp

```diff
@@ -0,0 +1,304 @@
+
+    vkWaitForFences(device, 1, &_inFlightFences[_currentFrame], VK_TRUE, UINT64_MAX);
+
+    uint32_t imageIndex;
+    VkResult result = vkAcquireNextImageKHR(device, _swapChain.get(), UINT64_MAX,
+                                            _imageAvailableSemaphores[_currentFrame], VK_NULL_HANDLE, &imageIndex);
+
+    if (result == VK_ERROR_OUT_OF_DATE_KHR)
+        return Result::Failure;
+    else if (result != VK_SUCCESS && result != VK_SUBOPTIMAL_KHR)
+        throw std::runtime_error("failed to acquire swap chain image!");
+
+    vkResetFences(device, 1, &_inFlightFences[_currentFrame]);
+
+    Command::RecordInfo recordInfo{};
+    recordInfo.currentFrame = _currentFrame;
+    recordInfo.imageIndex = imageIndex;
+    recordInfo.renderPass = _renderPass.get();
+    recordInfo.swapChainExtent = _swapChain.getExtent();
+    recordInfo.swapChainFramebuffers = _framebuffer.getSwapChainFramebuffers();
+    recordInfo.graphicsPipeline = _graphicsPipeline.get();
+
+    _command.recordBuffer(recordInfo);
+
+    VkSubmitInfo submitInfo{};
+    submitInfo.sType = VK_STRUCTURE_TYPE_SUBMIT_INFO;
+
+    VkSemaphore waitSemaphores[] = {_imageAvailableSemaphores[_currentFrame]};
+    VkPipelineStageFlags waitStages[] = {VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT};
+    submitInfo.waitSemaphoreCount = 1;
+    submitInfo.pWaitSemaphores = waitSemaphores;
+    submitInfo.pWaitDstStageMask = waitStages;
+
+    submitInfo.commandBufferCount = 1;
+    submitInfo.pCommandBuffers = &_command.getCommandBuffer(_currentFrame);
+
+    VkSemaphore signalSemaphores[] = {_renderFinishedSemaphores[_currentFrame]};
+    submitInfo.signalSemaphoreCount = 1;
+    submitInfo.pSignalSemaphores = signalSemaphores;
+
+    if (vkQueueSubmit(_logicalDevice.getGraphicsQueue(), 1, &submitInfo, _inFlightFences[_currentFrame]) != VK_SUCCESS)
+        throw std::runtime_error("failed to submit draw command buffer!");
+
+    VkPresentInfoKHR presentInfo{};
+    presentInfo.sType = VK_STRUCTURE_TYPE_PRESENT_INFO_KHR;
+    presentInfo.waitSemaphoreCount = 1;
+    presentInfo.pWaitSemaphores = signalSemaphores;
+
+    VkSwapchainKHR swapChains[] = {_swapChain.get()};
+    presentInfo.swapchainCount = 1;
+    presentInfo.pSwapchains = swapChains;
+    presentInfo.pImageIndices = &imageIndex;
+
+    result = vkQueuePresentKHR(_logicalDevice.getPresentQueue(), &presentInfo);
+
+    vkQueueWaitIdle(_logicalDevice.getPresentQueue());
+
+#if (MAX_FRAMES_IN_FLIGHT & (MAX_FRAMES_IN_FLIGHT - 1)) == 0
+    _currentFrame = (_currentFrame + 1) & (MAX_FRAMES_IN_FLIGHT - 1);
+#else
+    _currentFrame = _currentFrame + 1;
+    _currentFrame *= _currentFrame < MAX_FRAMES_IN_FLIGHT;
+#endif
+
+    if (result == VK_ERROR_OUT_OF_DATE_KHR || result == VK_SUBOPTIMAL_KHR || _framebufferResized)
+        return Result::Failure;
+    else if (result != VK_SUCCESS)
+        throw std::runtime_error("failed to present swap chain image!");
+
+    return Result::Success;
```

Comment:
I prefer not to factor, as this is a critical point in the program, which is called at each frame and handles internal parallelism in the Vulkan api. The code isn't particularly big in itself, but it gives the illusion that it's big because the structures it fills require a lot of parameters. But this function only calls 7 functions, which isn't much.

Reply:
👍🏼